src/newrelic/processor.go

```diff
@@ -663,42 +707,93 @@ func harvestByType(ah *AppHarvest, args *harvestArgs, ht HarvestType) {
 
 		customEvents := harvest.CustomEvents
 		harvest.CustomEvents = NewCustomEvents(eventConfigs.CustomEventConfig.Limit)
-		considerHarvestPayload(customEvents, args)
+		considerHarvestPayload(customEvents, args, duc)
 	}
 
 	if ht&HarvestErrorEvents == HarvestErrorEvents && eventConfigs.ErrorEventConfig.Limit != 0 {
 		log.Debugf("harvesting error events")
 
 		errorEvents := harvest.ErrorEvents
 		harvest.ErrorEvents = NewErrorEvents(eventConfigs.ErrorEventConfig.Limit)
-		considerHarvestPayload(errorEvents, args)
+		considerHarvestPayload(errorEvents, args, duc)
 	}
 
 	if ht&HarvestTxnEvents == HarvestTxnEvents && eventConfigs.AnalyticEventConfig.Limit != 0 {
 		log.Debugf("harvesting transaction events")
 
 		txnEvents := harvest.TxnEvents
 		harvest.TxnEvents = NewTxnEvents(eventConfigs.AnalyticEventConfig.Limit)
-		considerHarvestPayloadTxnEvents(txnEvents, args)
+		considerHarvestPayloadTxnEvents(txnEvents, args, duc)
 	}
 
 	if ht&HarvestSpanEvents == HarvestSpanEvents && eventConfigs.SpanEventConfig.Limit != 0 {
 		log.Debugf("harvesting span events")
 
 		spanEvents := harvest.SpanEvents
 		harvest.SpanEvents = NewSpanEvents(eventConfigs.SpanEventConfig.Limit)
-		considerHarvestPayload(spanEvents, args)
-
+		considerHarvestPayload(spanEvents, args, duc)
 	}
 
 	if ht&HarvestLogEvents == HarvestLogEvents && eventConfigs.LogEventConfig.Limit != 0 {
 		log.Debugf("harvesting log events")
 
 		logEvents := harvest.LogEvents
 		harvest.LogEvents = NewLogEvents(eventConfigs.LogEventConfig.Limit)
-		considerHarvestPayload(logEvents, args)
+		considerHarvestPayload(logEvents, args, duc)
+
+	}
+	// Only harvest data usage metrics if metrics are being harvested
+	if ht&HarvestDefaultData == HarvestDefaultData {
+		if args.blocking {
+			harvestDataUsage(args, duc)
+		} else {
+			go harvestDataUsage(args, duc)
+		}
+	}
+}
+
+func harvestDataUsage(args *harvestArgs, duc dataUsageController) {
+	duc.wg.Wait()
```

Comment:
Could this ever block forever?  Is there a wait to have a timeout?

Reply:
This theoretically could potentially block forever, resulting in a hung child thread. This could **only** happen in the instance that there is already another hung child thread, otherwise the wait group would be empty. And because the other threads that are blocking the wait group are just http calls that do have a timeout, in practice, this should never be hung.